src/SocialAuthUserManager.php

```diff
@@ -139,36 +142,71 @@ public function setSessionKeysToNullify(array $session_keys) {
   /**
    * Creates and/or authenticates an user.
    *
-   * @param string $email
-   *   The user's email address.
    * @param string $name
-   *   The user's name.
-   * @param string $id
-   *   The user's id in provider.
+   * The user's name.
+   * @param string $email
+   * The user's email address.
+   * @param string $type
+   * The social implementer identifier.
+   * @param string $social_media_id
+   * The unique id returned by the user.
    * @param string|bool $picture_url
-   *   The user's picture.
+   * The user's picture.
    *
    * @return \Symfony\Component\HttpFoundation\RedirectResponse
-   *   A redirect response.
+   * A redirect response.
    */
-  public function authenticateUser($email, $name, $id = NULL, $picture_url = FALSE) {
-    // Tries to load the user by their email.
-    $drupal_user = $this->loadUserByProperty('mail', $email);
-    // If user email has already an account in the site.
-    if ($drupal_user) {
-      // Authenticates and redirect existing user.
-      return $this->authenticateExistingUser($drupal_user);
+  public function authenticateUser($name, $email, $type, $social_media_id, $picture_url = FALSE) {
+
+    // Get ID of logged in user.
+    $uid = \Drupal::currentUser()->id();
+
+    if($uid){
+      return $this->addUserRecord($uid(), $type, $social_media_id);
+    }
+
+    // Checks for record in social _auth entity.
+    $drupal_user_id = $this->checkIfUserExists($type, $social_media_id);
+
+
+    if($drupal_user_id) {
+      // Load the user by their user_id.
+      $drupal_user = $this->loadUserByProperty('uid', $drupal_user_id);
+
+      // If user email has already an account in the site.
+      if ($drupal_user) {
+        // Authenticates and redirect existing user.
+        return $this->authenticateExistingUser($drupal_user);
+      }
+    }
+
+    if($email) {
+      // Load user by email.
+      $same_email_drupal_user = $this->loadUserByProperty('mail', $email);
+
+      //Check if User with same email account exists.
+
+      if ($same_email_drupal_user) {
+        //Add record for the same user.
+        $this->addUserRecord($same_email_drupal_user->id(), $type, $social_media_id);
+
+        // Authenticates and redirect the user.
+        return $this->authenticateExistingUser($same_email_drupal_user);
+      }
     }
 
-    $drupal_user = $this->createUser($name, $email);
-    // If the new user could be registered.
-    if ($drupal_user) {
+    $drupal_new_user = $this->createUser($name, $email);
```

Comment:
1) Add a comment about when this code would be executed. For instance: If user was not already logged in or registered, create new user.
2) Change variable to $drupal_user